<a href="https://colab.research.google.com/github/MuthiahAinun/Proyek_Analisis_Sentimen/blob/main/Skema3_Proyek_Analisis_Sentimen_Tsamarah_Muthi'ah_A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyek Analisis Sentimen: [Youtube-Comments-dataset]
- **Nama:** [Tsamarah Muthi'ah Abdullah]
- **Email:** [a135xaf486@devacademy.id]
- **ID Dicoding:** [a135xaf48]

# Import Library

In [ ]:
# Menginstall google API
!pip install --upgrade google-api-python-client pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.9/89.9 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 75.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 78.4 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.2
    Uninstalling pandas-2.2.2:
      Successfully uninstalled pandas-2.2.2
  Attempting uninstall: google-api-python-client
    Found existing installation: google-api-python-client 2.160.0
    Uninstalling google-api-python-client-2.160.0:
      Successfully uninstalled google-api-python-client-2.160.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.2.3 which is incompatible.


In [ ]:
!pip install gensim

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 2.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.7/26.7 MB 26.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 39.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 11.5 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.0.2
    Uninstalling numpy-2.0.2:
      Successfully uninstalled numpy-2.0.2
  Attempting uninstall: scipy
    Found existing installation: scipy 1.14.1
    Uninstalling scipy-1.14.1:
      Successfully uninstalled scipy-1.14.1


In [2]:
!pip install tensorflow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 644.9/644.9 MB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 980.2 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.5/24.5 MB 70.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 103.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 106.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 108.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 224.5/224.5 kB 16.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.5/72.5 kB 5.7 MB/s eta 0:00:00


In [4]:
!pip install imbalanced-learn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 238.4/238.4 kB 4.0 MB/s eta 0:00:00


In [1]:
# Import Library
import pandas as pd
import time
import random
import re
import string
import tensorflow as tf
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import seaborn as sns
import matplotlib.pyplot as plt
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, GRU, Conv1D, GlobalMaxPooling1D, Dense, Dropout, Bidirectional, BatchNormalization
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from sklearn.feature_extraction.text import TfidfVectorizer

# Data Preprocessing dan Pembersihan

In [2]:
# Membaca dan Menampilkan informasi dataset
dataset = pd.read_csv("NDsO1LT_0lw_youtube_comments.csv")
print("Jumlah data:", len(dataset))

Jumlah data: 85028


In [3]:
print(dataset.head())

             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  
0  54600  2025-02-08T16:59:31Z  
1      0  2025-03-19T21:51:42Z  
2      0  2025-03-19T21:43:03Z  
3      0  2025-03-19T21:34:32Z  
4      0  2025-03-19T21:20:21Z  


In [4]:
print("Informasi Dataset:")
print(dataset.info())

Informasi Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85028 entries, 0 to 85027
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   username      84982 non-null  object
 1   text          85028 non-null  object
 2   likes         85028 non-null  int64 
 3   published_at  85028 non-null  object
dtypes: int64(1), object(3)
memory usage: 2.6+ MB
None


In [5]:
print('\nStatistik Deskriptif:')
print(dataset.describe())


Statistik Deskriptif:
               likes
count   85028.000000
mean        9.727654
std       744.702565
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max    167348.000000


In [6]:
# Mengecek jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        46
text             0
likes            0
published_at     0
dtype: int64


In [7]:
# Menghapus nilai kosong
dataset.dropna(inplace=True)

In [8]:
# Menghapus duplikasi
dataset.drop_duplicates(subset=['text'], inplace=True)

In [9]:
# Pembersihan teks
def clean_text(text):
    text = re.sub(r'@\w+', '', text)  # Menghapus username
    text = re.sub(r'http\S+', '', text)  # Menghapus URL
    text = re.sub(r'[^A-Za-z\s]', '', text)  # Menghapus karakter non-huruf
    text = text.lower().strip()  # Mengubah ke huruf kecil dan menghapus spasi
    return text

dataset['cleaned_text'] = dataset['text'].apply(clean_text)
print("Data setelah pembersihan:")
print(dataset.head())

Data setelah pembersihan:
             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  \
0  54600  2025-02-08T16:59:31Z   
1      0  2025-03-19T21:51:42Z   
2      0  2025-03-19T21:43:03Z   
3      0  2025-03-19T21:34:32Z   
4      0  2025-03-19T21:20:21Z   

                                        cleaned_text  
0  beast games finale drops february th go watch ...  
1                               saya orang indonesia  
2                                                     
3                                                     
4  this is for girls listen to dalida

In [10]:
# Mengecek kembali jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        0
text            0
likes           0
published_at    0
cleaned_text    0
dtype: int64


# **2. Ekstraksi Fitur dan Pelabelan Data**

Metode yang digunakan bebas sesuai dengan preferensi masing-masing peserta. Tahapan ini penting untuk mempersiapkan data sehingga dapat diolah lebih lanjut dalam proses pelatihan model.

In [36]:
# Melabeli komentar secara manual (positif, netral, negatif)
def label_comment(text):
    if any(word in text for word in ["love", "great", "awesome", "good", "nice", "amazing", "fantastic"]):
        return "positif"
    elif any(word in text for word in ["hate", "bad", "awful", "worst", "terrible", "disgusting", "boring"]):
        return "negatif"
    else:
        return "netral"

dataset['label'] = dataset['cleaned_text'].apply(label_comment)
print("Distribusi label:")
print(dataset['label'].value_counts())

Distribusi label:
label
netral     63770
positif     4724
negatif       69
Name: count, dtype: int64


In [66]:
# Encode label dengan LabelEncoder
le = LabelEncoder()
dataset['label'] = le.fit_transform(dataset['label'])

# **3. Pembangunan Model Deep Learning**

Pilihan algoritma pelatihan ini haruslah sesuai dengan tujuan analisis sentimen yang ingin dicapai.

# Skema 3: GRU dengan Word2Vec (70/30)

**Algoritma:** GRU (Gated Recurrent Unit)

**Ekstraksi Fitur:** Word2Vec

**Alasan Pemilihan:**

- GRU lebih sederhana dan lebih ringan dibandingkan dengan LSTM, sehingga lebih cepat dalam proses pelatihan tanpa kehilangan banyak informasi temporal. Cocok untuk dataset teks dengan jumlah data besar.
- Word2Vec digunakan untuk menghasilkan vektor kata yang memiliki makna semantik sehingga model dapat memahami konteks dari kata-kata dalam kalimat.

In [67]:
print("\nSkema 3: GRU dengan Word2Vec (70/30)")
# Split data 70/30
X = dataset['cleaned_text']
y = dataset['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


Skema 3: GRU dengan Word2Vec (70/30)


In [68]:
# Tokenizer dan Padding
tokenizer = Tokenizer(num_words=10000, oov_token="<OOV>")
tokenizer.fit_on_texts(X_train)  # Hanya fit pada data latih untuk mencegah data bocor
X_train_sequences = tokenizer.texts_to_sequences(X_train)
X_test_sequences = tokenizer.texts_to_sequences(X_test)

# Padding
X_train_padded = pad_sequences(X_train_sequences, maxlen=209, padding='post')
X_test_padded = pad_sequences(X_test_sequences, maxlen=209, padding='post')

In [69]:
# Mengatasi Ketidakseimbangan Data dengan SMOTE
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_train_padded, y_train)

print("Distribusi label setelah SMOTE:")
print(pd.Series(y_resampled).value_counts())

Distribusi label setelah SMOTE:
label
1    44619
2    44619
0    44619
Name: count, dtype: int64


In [70]:
# One-Hot Encoding pada label
encoder = LabelEncoder()
y_resampled_encoded = encoder.fit_transform(y_resampled)  # Transform label SMOTE
y_test_encoded = encoder.transform(y_test)  # Transform label uji


In [71]:
# One-Hot Encoding
y_resampled_encoded = tf.keras.utils.to_categorical(y_resampled_encoded, num_classes=3)
y_test_encoded = tf.keras.utils.to_categorical(y_test_encoded, num_classes=3)

print("Label Setelah One-Hot Encoding:")
print(y_resampled_encoded[:5])
print("Classes:", encoder.classes_)

Label Setelah One-Hot Encoding:
[[0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]]
Classes: [0 1 2]


In [75]:
# Model GRU
model3 = Sequential([
    Embedding(input_dim=10000, output_dim=64, input_length=209),  # Embedding layer lebih kecil
    Bidirectional(GRU(64, return_sequences=True, dropout=0.2, recurrent_dropout=0.2)),  # GRU lebih kecil
    GlobalMaxPooling1D(),  # Pooling
    BatchNormalization(),  # Normalisasi batch
    Dropout(0.3),  # Dropout lebih rendah
    Dense(64, activation='relu'),  # Layer Dense lebih kecil
    Dropout(0.3),  # Dropout lebih rendah
    Dense(32, activation='relu'),  # Layer Dense lebih kecil
    Dropout(0.3),  # Dropout lagi
    Dense(3, activation='softmax')  # Output layer dengan 3 kelas
])

In [76]:
model3.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

In [79]:
# Melatih model dengan data seimbang
history = model3.fit(
    X_resampled, y_resampled_encoded,  # Data dan label sudah One-Hot
    epochs=10,
    validation_data=(X_test_padded, y_test_encoded),  # Data validasi juga One-Hot
    batch_size=64
)

Epoch 1/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1219s 583ms/step - accuracy: 0.9227 - loss: 0.2071 - val_accuracy: 0.9725 - val_loss: 0.1052
Epoch 2/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1227s 585ms/step - accuracy: 0.9344 - loss: 0.1814 - val_accuracy: 0.9698 - val_loss: 0.1081
Epoch 3/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1205s 576ms/step - accuracy: 0.9424 - loss: 0.1597 - val_accuracy: 0.9700 - val_loss: 0.1204
Epoch 4/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1202s 575ms/step - accuracy: 0.9476 - loss: 0.1458 - val_accuracy: 0.9578 - val_loss: 0.1437
Epoch 5/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1219s 573ms/step - accuracy: 0.9534 - loss: 0.1327 - val_accuracy: 0.9736 - val_loss: 0.1010
Epoch 6/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1254s 589ms/step - accuracy: 0.9583 - loss: 0.1194 - val_accuracy: 0.9632 - val_loss: 0.1492
Epoch 7/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1259s 578ms/step - accuracy: 0.9602 - loss: 0.1150 - val_accuracy: 0.9736 - val_loss: 0.1047
Epoch 8/10
2092/2092 ━━━━━━━━━━━━━━━━━━━━ 1197s 566ms/s

In [81]:
# Preprocessing data testing
X_test_sequences = tokenizer.texts_to_sequences(X_test)
X_test_padded = pad_sequences(X_test_sequences, maxlen=209, padding='post')

# Evaluasi model pada data testing
test_loss, test_accuracy = model3.evaluate(X_test_padded, y_test_encoded)
print(f"Akurasi Testing (dari model.evaluate): {test_accuracy:.2%}")

643/643 ━━━━━━━━━━━━━━━━━━━━ 52s 80ms/step - accuracy: 0.9723 - loss: 0.1182
Akurasi Testing (dari model.evaluate): 97.05%


In [84]:
target_names = [str(cls) for cls in le.classes_]
print(target_names)  # Pastikan hasilnya dalam bentuk list of strings

['0', '1', '2']


In [83]:
print(type(le.classes_))
print(le.classes_)
print(list(le.classes_))

<class 'numpy.ndarray'>
[0 1 2]
[np.int64(0), np.int64(1), np.int64(2)]


In [85]:
# Prediksi pada data testing
y_pred2 = np.argmax(model3.predict(X_test_padded), axis=1)
y_true = np.argmax(y_test_encoded, axis=1)

print(classification_report(y_true, y_pred2, target_names=target_names))

643/643 ━━━━━━━━━━━━━━━━━━━━ 52s 81ms/step
              precision    recall  f1-score   support

           0       0.02      0.42      0.04        12
           1       1.00      0.97      0.98     19151
           2       0.80      0.97      0.87      1406

    accuracy                           0.97     20569
   macro avg       0.61      0.78      0.63     20569
weighted avg       0.98      0.97      0.98     20569



In [86]:
# Inference  Skema 3
print("\nInference Skema 3: GRU dengan Word2Vec (70/30)")
def predict_text(text, tokenizer, model3, le, max_length=100):
    # Preprocessing teks
    text = text.lower()
    text_seq = tokenizer.texts_to_sequences([text])
    text_pad = pad_sequences(text_seq, maxlen=max_length, padding='post')

    # Prediksi
    prediction = model3.predict(text_pad)
    predicted_label = np.argmax(prediction, axis=1)[0]
    predicted_class = le.inverse_transform([predicted_label])[0]

    print(f"Teks: {text}")
    print(f"Prediksi Label: {predicted_class}")
    print(f"Confidence: {np.max(prediction) * 100:.2f}%")
    return predicted_class


Inference Skema 3: GRU dengan Word2Vec (70/30)


In [91]:
# Melakukan Prediksi pada Teks Baru (0: Negatif, 1: Netral, 2: Positif)
predict_text("I love how you guys taking the video, that was amazing!", tokenizer, model3, le)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 62ms/step
Teks: i love how you guys taking the video, that was amazing!
Prediksi Label: 2
Confidence: 100.00%


np.int64(2)

**Insight Skema 3:**

1. Model memiliki akurasi tinggi sebesar 97.05%, yang menunjukkan bahwa model secara keseluruhan cukup andal dalam mengklasifikasikan data.
2. **Kelas 0 (Negatif):**

- Precision: 0.02 (Sangat rendah)

- Recall: 0.42 (Cukup tinggi untuk jumlah data yang sedikit)

- F1-Score: 0.04 (Buruk)

- Jumlah data: 12 (Sangat sedikit)

- **Catatan:** Model mengalami kesulitan dalam mengklasifikasikan kelas ini karena jumlah sampel yang sangat sedikit sehingga model cenderung mengabaikannya.

3. **Kelas 1 (Netral):**

- Precision: 1.00 (Sangat baik)

- Recall: 0.97 (Sangat baik)

- F1-Score: 0.98 (Sangat tinggi)

- Jumlah data: 19,151 (Mayoritas data)

- **Catatan:** Model sangat akurat dalam mendeteksi kelas ini karena data yang melimpah

4. **Kelas 2 (Positif):**

- Precision: 0.80 (Baik)

- Recall: 0.97 (Sangat baik)

- F1-Score: 0.87 (Tinggi)

- Jumlah data: 1,406 (Cukup banyak)

- **Catatan**: Model juga mampu menangani kelas positif dengan cukup baik.

5. Model memprediksi teks "i love how you guys taking the video, that was amazing!" sebagai kelas 2 (Positif) dengan 100% confidence.